### ISSUE TITLE
Old documentation at https://fishshell.com/docs/X.Y is confusing

### ISSUE THREAD

**Issue by andmis** [fish-shell/fish-shell#8503]
I have been using fish for a couple of years now (and have actually read most of the manual thoroughly) and I only just today realized that the documentation on those pages (e.g., https://fishshell.com/docs/2.2/) is for old versions of the shell. Sometimes old versions of the docs come up in Google searches (apparently). I suspect most users don't know the version number of the latest version of fish (I could not have said off the top of my head whether I was on 2.x or 3.x) so a URL like https://fishshell.com/docs/2.2/ does not immediately scream "out-of-date". And it isn't mentioned anywhere else in the docs that they are for an old version of fish.

It looks like old versions of the docs were mirrored deliberately, so presumably someone had a reason for doing so. But it confused me for a long time apparently, so that's [one data point](https://github.com/fish-shell/fish-shell/issues/8502).

**Comment by faho** [comment]
Yes, it's the old documentation. We keep it around for old versions and don't plan on making any changes to it.

The documentation linked from the main site is the current version.

**Comment by zanchey** [comment]
Some projects (eg CMake) put a large banner at the top of every page for the old version to say "This documents an old version of CMake. Click here to see the latest release.". We'd have to add it by hand, though - perhaps something to look at after 3.4.0?

**Comment by faho** [comment]
Yeah, that makes sense.

We should also look at removing very old documentation - pointing people at the 2.2.0 docs isn't helpful.

**Comment by zanchey** [comment]
I kind of like having the old versions accessible - sometimes saves spelunking in the git history

**Comment by andrey-mishchenko** [comment]
I did some investigation. Here's a summary so far.

# General comments

There are two things to think about: (1) adding banners to old versions of the docs to notify users that the docs are old, and to refer them to the latest version, and (2) where exactly the links will point -- to the root of the latest docs, or to the "same page" of the latest docs.

I tend to agree that it's nice to keep old versions of the docs around, and that it wouldn't be an issue to do so if they were labeled with a banner.

There are two systems that we've used to generate docs (since fish 2.0, which is the oldest version for which we have docs online): Doxygen and Sphinx. Doxygen was used for 2.0 through 3.0 inclusive, and Sphinx has been used since 3.1, through today. (We don't have docs online for patch versions, just 2.0 through 2.7 and 3.0 through today.)

# Adding a "this is an old version" banner

## Doxygen

As far as I can tell (after a bunch of digging online), Doxygen does not have any built-in mechanisms for adding an "old version" banner, or for multiple versions of a project.

**Proposal:** For the Doxygen-generated docs (fish v3.0 and below), we should just edit the existing, generated HTML/CSS by hand to insert a banner, in a one-off PR to `fish-site`.

The downside to this is that if, for some reason, someone comes along and re-generates the HTML/CSS for those old versions of fish, that banner could be lost. Hopefully in this case the diff view would show the manually-added banner in red and the author would realize they should readd it.

One alternative would be to push new patch versions (adding the banner somehow) to all old versions of `fish-shell` for which we want to keep docs on the website. This seems not to be worth it.

## Sphinx

TODO. Some notes:

The Python folks have some pretty-simple home-grown code implementing their deprecation banner:

<https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Doc/README.rst#deprecation-header>
<https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Doc/tools/templates/layout.html>

# Where do we actually link?

If we add a "this is an old version" banner, the banner should have (at least) a link to the latest version of the docs. Does this link point to the root of the latest docs, or to a URL constructed from the current URL attempting to approximate the page the user is on?

Some projects ([CMake](https://cmake.org/cmake/help/v3.5/), [Python](https://docs.python.org/3.9/)) also provide a dropdown that allows users to select different versions of (a particular page of) the docs. This dropdown navigates by page -- meaning if you are on the `print` documentation for Python v3.8 and you select v3.7 in the dropdown, you will go to the `print` documentation for v3.7.

Both CMake and Python use Sphinx, which is what we are using since 3.1. I investigated how they implement this functionality.

As far as I can tell, Sphinx does not have this feature built in. The Python and CMake folks have home-grown code in the build scripts for their docs, as well as custom JS on their websites implementing the URL translation for navigating among different versions of a particular page of documentation (e.g., for `print`). As far as I can tell they aren't using any packages or anything for this.

<https://github.com/python/docsbuild-scripts/blob/328a5c2e4d7d7baad7f0d7384c69765a5c5fdd49/templates/switchers.js>
<https://github.com/python/docsbuild-scripts/blob/328a5c2e4d7d7baad7f0d7384c69765a5c5fdd49/build_docs.py#L354>
<https://gitlab.kitware.com/cmake/cmake/-/commit/82e5ab9e14b81e0d21ef61f503a2b4c8638776e8>
<https://cmake.org/cmake/help/version_switch.js>

I originally thought it would be a nightmare to implement but it seems like it might not be so bad. The structure (in the sense of URLs) of our docs changed substantially between Doxygen and Sphinx, but some things translate nicely, for example, you can go all the way from 2.0 to the current docs for many commands: <https://fishshell.com/docs/2.0/commands.html#alias> to <https://fishshell.com/docs/current/cmds/alias.html>.

**Comment by zanchey** [comment]
This is a great writeup, thanks!

**Comment by ridiculousfish** [comment]
The Python approach requires us to be able to build old docs from new versions of fish - I don't think we want to put ourselves on the hook for that (e.g. recall the Doxygen switch).

I manually added old-docs warnings to previous docs. For 3.5+, we have a new system: the site requests `/release_version.json` and shows the banner if that is newer than what is baked into the docs site. `release_version.json` is updated automatically when the site is built (it's just a Liquid template) so this doesn't add any additional steps.

Anyways an example is [here](https://fishshell.com/docs/3.4/index.html). Closing, thanks for filing.
